src/Features/Core/Portable/PullMemberUp/CodeActionAndSolutionGenerator.cs

```diff
@@ -0,0 +1,197 @@
+            ImmutableDictionary<ISymbol, IEnumerable<SyntaxNode>> syntaxMap)
+        {
+            foreach (var analysisResult in result._membersAnalysisResults)
+            {
+                foreach (var syntax in syntaxMap[analysisResult._member])
+                {
+                    RemoveDeclaration(editorMap[solution.GetDocumentId(syntax.SyntaxTree)], syntax);
+                }
+            }
+        }
+
+        private async Task<(ImmutableDictionary<DocumentId, DocumentEditor>, ImmutableDictionary<ISymbol, IEnumerable<SyntaxNode>>)> InitializeEditorMapsAndSyntaxMapAsync(
+            AnalysisResult result,
+            SyntaxNode targetSyntaxNode,
+            SolutionEditor solutionEditor,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            // Members and target may come from different documents,
+            // So EditorMap is used to save and group all the editors will be used.
+            var editorMapBuilder = ImmutableDictionary.CreateBuilder<DocumentId, DocumentEditor>();
+            // One member may have multiple syntaxNodes (e.g partial method).
+            // SyntaxMap is used to find the syntaxNodes need to be changed more easily.
+            var syntaxMapBuilder = ImmutableDictionary.CreateBuilder<ISymbol, IEnumerable<SyntaxNode>>();
+            var membersNeedToBeChanged = result._membersAnalysisResults;
+
+            foreach (var memberAnalysisResult in membersNeedToBeChanged)
+            {
+                var tasks = memberAnalysisResult._member.DeclaringSyntaxReferences.Select(@ref => @ref.GetSyntaxAsync(cancellationToken));
+                var allSyntaxes = await Task.WhenAll(tasks).ConfigureAwait(false);
+                await AddEditorsToEditorMapBuilderAsync(editorMapBuilder, allSyntaxes, solutionEditor, solution);
+                syntaxMapBuilder.Add(memberAnalysisResult._member, allSyntaxes.ToImmutableArray());
+            }
+
+            await AddEditorsToEditorMapBuilderAsync(editorMapBuilder, new SyntaxNode[] { targetSyntaxNode }, solutionEditor, solution);
+            return (editorMapBuilder.ToImmutableDictionary(), syntaxMapBuilder.ToImmutableDictionary());
+        }
+
+        private async Task AddEditorsToEditorMapBuilderAsync(
+            ImmutableDictionary<DocumentId, DocumentEditor>.Builder mapBuilder,
+            IEnumerable<SyntaxNode> syntaxNodes,
+            SolutionEditor solutionEditor,
+            Solution solution)
+        {
+            foreach (var syntax in syntaxNodes)
+            {
+                if (!mapBuilder.ContainsKey(solution.GetDocumentId(syntax.SyntaxTree)))
+                {
+                    var id = solution.GetDocumentId(syntax.SyntaxTree);
+                    var editor = await solutionEditor.GetDocumentEditorAsync(id).ConfigureAwait(false);
+                    mapBuilder.Add(solution.GetDocumentId(syntax.SyntaxTree), editor);
```

Comment:
everywhere i've looked, i see you doing `solution.GetDocumentId(syntax.SyntaxTree)`.  Why not just have a mapping from SyntaxTree->Editor?  Do you need to be keyed off of DOcumentId?